### PR TITLE
Completion improvements

### PR DESCRIPTION
**Pull request by paulvanbrenk**
This adds a WithDescription method on the CompletionItem, this makes it easy to add synchronously created completion descriptions to a completion item. e.g. in the case of Snippets.

@CyrusNajmabadi, @mattwar

### CODE REVIEW

File: src/Features/Core/Portable/Completion/CompletionItem.cs [hunk 1]
```diff
@@ -196,6 +196,42 @@ public CompletionItem WithTags(ImmutableArray<string> tags)
...
+            var descriptionTags = ImmutableArray.Create(new TaggedText(TextTags.Text, description));
+
+            return WithDescription(descriptionTags);
+        }
+
+        public CompletionItem WithDescription(ImmutableArray<TaggedText> description)
+        {
+            var properties = Properties ?? ImmutableDictionary<string, string>.Empty;
+
+            var encoded = CommonCompletionUtilities.EncodeDescription(description);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItem.cs]: consider inlining these locals.

**Comment by CyrusNajmabadi** [comment]
Nice.  I like this.  @mattwar  what do you think?

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by mattwar** [comment]
No this API should not exist on CompletionItem. 

CompletionItem's are not supposed to have descriptions as part of their contract. That's why there is a separate API on the CompletionService for getting the description.  There's nothing wrong with actually attaching known descriptions to an item and then returning this value from the GetDescriptionAsync API, but when you add it to the completion item API it confuses the next person that comes along and implements a provider or service or host.  Do all completion items have descriptions?  Why would I need to call this separate API if the item has a method to get it directly?

 This helper API already exists as a static function on CommonCompletionItem but it is internal. Consider exposing this as a helper on a separate type. Maybe exposing CommonCompletionItem itself (thought that is a bigger pill to swallow), or maybe a new type.

**Comment by paulvanbrenk** [comment]
@mattwar How about making this an extension method?

**Comment by mattwar** [comment]
@paulvanbrenk I understand your desire to make this helper easily discoverable, but by doing that you've changed the apparent contract of what a completion item is.

**Comment by paulvanbrenk** [comment]
Ok, I'll add a CompletionItemHelper class, to add an encoded description the properties collection.

**Comment by CyrusNajmabadi** [comment]
One problem with teh current model Matt is that Descriptions effectively become part of our internal contract.  i.e. if someone adds a Property called "Description" and it's not encoded as we expect, then crashes can happen.

@paulvanbrenk my recommendation would be to add helpers to Encode/Decode on the TaggedText type.  These helpers would also be there for lists of TaggedTexts as well.  That will make adding/retrieving the description much easier.

> CompletionItem's are not supposed to have descriptions as part of their contract. That's why there is a separate API on the CompletionService for getting the description. There's nothing wrong with actually attaching known descriptions to an item and then returning this value from the GetDescriptionAsync API, 

Matt, this is what Paul is trying to do.  However, the API currently on CompletionServiceWithProviders works htis way:

```c#
        public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (CommonCompletionItem.HasDescription(item))
            {
                return Task.FromResult(CommonCompletionItem.GetDescription(item));
            }

            var provider = GetProvider(item);
            if (provider != null)
            {
                return provider.GetDescriptionAsync(document, item, cancellationToken);
            }
            else
            {
                return Task.FromResult(CompletionDescription.Empty);
            }
        }
```

Basically, as part of the *standard* contract we're exposing, we're stating that we will check the item itself for the description.  However, currently there is no way to actually put a description on an item.  So our standard impl does something that an actual 3rd party can't effectively interact with.  This PR changes that so that this codepath is available to 3rd parties like TS.

So, IMO, we either expose these, or we change this implementation of this method to not perform this shotcircuiting.  Instead, we can just make the impl defer to the provider, and we can make the base C#/VB provider pull the data off the item.  TS can do the same for their items.

**Comment by mattwar** [comment]
@CyrusNajmabadi I agree that the fact that the CompletionServiceWithProviders class checks this property is a bug.

**Comment by CyrusNajmabadi** [comment]
Ok.  If it moves to common completion service, then there's still the open question on how someone should imbue the item with this data.  Having it be a helper method versus an instance method doesn't seem to make much difference at all afaict.

**Comment by mattwar** [comment]
@CyrusNajmabadi making it an instance method makes it description appear as an expected part of the all CompletionItems, which it is not. If you have a separate helper API that lets you annotate a CompletionItem with a description then it would generally be assumed that CompletionItem's would not normally have this annotation.

**Comment by CyrusNajmabadi** [comment]
How about an extension method?

**Comment by mattwar** [comment]
@CyrusNajmabadi and extension method just makes it appear like a part of the CompletionItem API which is the point of my objection.

**Comment by CyrusNajmabadi** [comment]
I confused CommonCompletionService and CompletionService.  I agree now with what Matt said.  We should move this specialized logic to CommonCompletionService.  The core layer shuld have no special processing around this area.

**Comment by CyrusNajmabadi** [comment]
@mattwar is going to do some refactoring here.

**Comment by paulvanbrenk** [comment]
great, can you add me to the pr?

**Comment by paulvanbrenk** [comment]
Is there anything we want to keep from this PR?

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!
